src/main/java/com/aws/greengrass/authorization/AuthorizationModule.java

```diff
@@ -42,17 +47,61 @@ public void addPermission(String destination, Permission permission) throws Auth
                 || Utils.isEmpty(permission.getOperation())) {
             throw new AuthorizationException("Invalid arguments");
         }
-        // resource as null is ok, but it should not be empty
         String resource = permission.getResource();
-        if (resource != null && Utils.isEmpty(resource)) {
-            throw new AuthorizationException("Resource cannot be empty");
-        }
+        String validatedResource = validateResource(resource);
         resourceAuthZCompleteMap.get(destination).get(permission.getPrincipal()).get(permission.getOperation()).add(
-                permission.getResource());
+                resource);
         rawResourceList.get(destination).get(permission.getPrincipal()).get(permission.getOperation()).add(
-                permission.getResource());
+                validatedResource);
+    }
+
+    /**
+     * Only allow '?' if it's escaped. You can only escape special characters ('*', '$', '?').
+     * Any occurrence of '${' is only valid if it holds a single valid special character ('*', '$', '?') inside it
+     * and ends with '}'. (eg: "${*}" is valid, "${c}" is invalid, "${c" is invalid, ${*bc} is invalid)
+     * @param resource resource to be parsed
+     */
+    private String validateResource(String resource) throws AuthorizationException {
+        if (resource == null) {
+            return null;
+        }
+        // resource as null is ok, but it should not be empty
+        if (Utils.isEmpty(resource)) {
+            throw new AuthorizationException("Resource cannot be empty");
+        }
+        int length = resource.length();
+        StringBuilder sb = new StringBuilder(length);
+        for (int i = 0; i < length; i++) {
+            char currentChar = resource.charAt(i);
+            if (currentChar == escapeChar && i + 1 < length && resource.charAt(i + 1) == '{') {
+                char actualChar = getActualChar(resource.substring(i));
+                if (actualChar == nullChar) {
+                    throw new AuthorizationException("Resource not allowed, incorrect escape sequence used");
+                }
+                if (!isSpecialChar(actualChar)) {
+                    throw new AuthorizationException("Resource not allowed, Only special characters "
```

Comment:
nit: only

Reply:
I also wonder if those exception should say"Invalid arguments" instead of "Resource not allowed"

Reply:
For Customers, its the policy they are adding thats failing. I think this tells why it is failing.

Reply:
I'm not sure about throwing here. Logging, sure, but not about throwing. 

I do not like how we handle auth policy issues right now in that it can prevent ALL policies from working, not only the broken policies.

Reply:
I had talked to Joe about this, I think he said we want to fail as early as possible maybe at deployment so the user knows the problem . I could have understood it wrong. Joe hasn't reviewed this and said he will look at this on Monday. We can discuss it more then. But I agree this prevents all policies from working.

Reply:
It doesn't cause the deployment to fail, so I don't think this is good.

Reply:
I think this validation is in addPermission() which is called when loading/updating configs. So I think the log should point out that they need to fix that resource argument but I just feel "Resource not allowed" sounds like it's not authorized..

Reply:
Oh okay, so  what will happen when this error is thrown, that particular merge-config just won't work?

Reply:
Trace the calls. Who calls this and what do they do with the exception?

Reply:
> I think this validation is in addPermission() which is called when loading/updating configs. So I think the log should point out that they need to fix that resource argument but I just feel "Resource not allowed" sounds like it's not authorized..
hmm...makes sense @youtuyy . Any suggestions, like "Invalid resource in AccessControl, resource contains incorrect escape sequence" or something?

Reply:
yeah probably just something like "Resource cannot contain invalid escape sequence + some details.. " also depending on the path the exception is propagated. You can also check with some integ tests what will be logged and thrown

Reply:
> Trace the calls. Who calls this and what do they do with the exception?

@MikeDombo The only time the policies are loaded is initially and when there's a merge/update config. The behavior right now is, any other resources included in that same policy(with an invalid resource) will not work and throw exception which will be logged by AuthorizationHandler. But a separate policy that is before/after the invalid policy will still be loaded and will work.
I can change it to reject a particular resource inside the policy, and continue with the next resource, so only that resource is rejected.